---
title: "RNAseq: count, filter, normalise"
author: "RWH"
date: "`r format(Sys.time(), '%d %B %Y')`"

output: 
  html_document:
    theme: yeti 
    highlight: pygments 
    anchor_sections: FALSE
---

```{r setup, echo = FALSE, message = FALSE, warning = FALSE}

knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = TRUE) 

library(tidyverse)
library(here)
# library(HTSFilter)
library(edgeR)

dir_scripts <- here(dir_root, "scripts")
dir_save <- here(dir_data, "RMD_output")

source(here(dir_scripts, "plotting_themes.R"))
source(here(dir_scripts, "plotting.R"))
source(here(dir_scripts, "analysis.R"))

```

```{r set_parameters}

# set as "mature", "stemloop" or "combined"
# analysis_scope <- "mature" 

# filtering
# to keep a gene, must be present above cpm threshold on n samples
# cpm_threshold <- 1        
# n_samples <- 8         

miR_spike_in <- "cel-miR-39"
norm_method <- "TMM" # TMM or spike

```

```{r import_counts, message = FALSE}

# first import raw data to generate a single counts .csv
# source(here(dir_data, "x1_import_counts.R"))

df_rawcounts <- read_csv(here(dir_save, "rawcounts.csv"))

```

```{r merge_stemloop, include = FALSE}

# df_rawcounts_mature %>% 
#   pivot_longer(cols = -gene, names_to = "library", values_to = "mature") -> df_m_L
# 
# df_m_L$gene <- tolower(df_m_L$gene)
# 
# df_rawcounts_stemloop %>% 
#   pivot_longer(cols = -gene, names_to = "library", values_to = "stemloop") -> df_sl_L
# 
# df_sl_L$gene <- tolower(df_sl_L$gene)
# 
# df_L <- full_join(df_m_L, df_sl_L, by = c("gene", "library"))
# df_L[is.na(df_L)] <- 0
# df_L %>% mutate(total = mature + stemloop) -> df_L
# 
# 
# if (analysis_scope == "mature") {
# 
# df_L %>% 
#   select(gene, library, mature) %>% 
#   pivot_wider(names_from = library, values_from = mature) -> df_rawcounts
# 
# }
# 
# if (analysis_scope == "stemloop") {
# 
# df_L %>% 
#   select(gene, library, stemloop) %>% 
#   pivot_wider(names_from = library, values_from = stemloop) -> df_rawcounts
# 
# }
# 
# if (analysis_scope == "combined") {
# 
# df_L %>% 
#   select(gene, library, total) %>% 
#   pivot_wider(names_from = library, values_from = total) -> df_rawcounts
# 
# }

```

```{r import_metadata}

df_samples <- read_csv(here(dir_data, "meta.csv"))

read_delim(here(dir_data, "counts", "lib_size.txt")) %>% 
  rename(lib.size = reads) -> df_libsize

df_libsize$sample <- df_libsize$library %>% gsub(".trim.fq", "", .)
left_join(df_samples, df_libsize) -> df_samples

source(here(dir_data, "x2_repair_names.R"))

```


## Overview of experimental design

```{r experimental_description, child=here(dir_data, "exp_setup.txt"), warning=FALSE}

```

<br>
<br>

### Sample metadata

```{r groups}

df_samples %>% knitr::kable()

```

<br>
<br>

## Analysis parameters  

This analysis looks at `r # analysis_scope` reads mapping to mature forms of miRNA.  

<!-- When filtering data, genes are retained if they are expressed at levels over `r # cpm_threshold` cpm in at least `r # n_samples` libraries.  -->

<br>
<br>

## Mapping

Extent of mapping.  Note that in this analysis, the denominator is the total number of reads in the trimmed libraries (i.e. before deduplication).  Therefore this will include duplicate UMIs and so will end up slightly under-estimating the proportion of deduplicated reads that are mapped.  

```{r count_mapping}

df_rawcounts %>% 
  pivot_longer(cols = c(-gene, -genome), names_to = "library", values_to = "count") %>% 
  group_by(library, genome) %>% 
  summarise(
    total_reads = sum(count)
  ) -> df_map_summary

df_libsize %>%
  select(sample, lib.size) %>% 
  left_join(df_map_summary, ., by = c("library" = "sample")) %>% 
  mutate(
    mapped_pc = (total_reads / lib.size)*100
    ) -> df_map_summary

df_map_summary$genome %>% 
  as.factor() %>% 
  fct_relevel("spike", "dog", "mouse", "human") %>% 
  fct_rev() -> df_map_summary$genome

df_samples_short <- df_samples %>% select(library = sample, group)
df_map_summary$library <- df_map_summary$library %>% as.factor() %>% fct_relevel(levels(df_samples$sample))

df_map_summary %>% 
  left_join(df_samples_short) %>% 
  ggplot(aes(x = library, y = mapped_pc, fill = genome)) +
    geom_col() +
    ylab("reads mapped (%)") +
    coord_flip() +
    theme_miR_vertical() +
    facet_wrap(~group) +
    sf_miR_3_groups

```


## Make DGL object

Construct DGL object for differential expression analysis in edgeR.  

```{r make_dgl, warning = FALSE}

# Construct DGEList object in edgeR
# 
# DGEList contains:  
# 
# - counts = matrix of counts
# - samples = df of sample information, to include group & lib.size  
# - genes (optional) = df of gene annotation features


# ensure rawcounts and sample orders are the same
libraries_in_order <- df_samples$sample %>% levels()
df_rawcounts <- df_rawcounts %>% select(gene, genome, libraries_in_order)

# convert rawcounts to named matrix
df_rawcounts_trim <- df_rawcounts %>% select(-genome)
df_rawcounts_trim[is.na(df_rawcounts_trim)] <- 0

genes <- df_rawcounts_trim$gene
df_rawcounts_trim <- df_rawcounts_trim %>% select(-gene)
rownames(df_rawcounts_trim) <- genes

rawcounts <- data.matrix(df_rawcounts_trim)


# remove spike-in
spike_in_index <- which(rownames(rawcounts) == miR_spike_in)
spike_in_factor <- as.numeric(rawcounts[spike_in_index, ] / colSums(rawcounts))
rawcounts_without_spikein <- rawcounts[rownames(rawcounts) != miR_spike_in, ]


# construct DGEList
dgl_raw <- DGEList(
  counts = rawcounts_without_spikein,
  samples = select(df_samples, -group, -lib.size),
  lib.size = df_samples$lib.size,                       # think this really affects the filtering step
  group = df_samples$group
)

```

There are `r nrow(dgl_raw)` genes in the raw dataset. 

Structure of sample data (showing first 6 samples):  

```{r print_dgl}
dgl_raw$samples %>% head() %>% knitr::kable()
```


<br>
<br>

## Filter & normalise

<!-- 

```{r get_cpm, eval = FALSE}

# NB can also collapse technical replicates at this point if applicable

counts_raw <- dgl_raw$counts %>% as_tibble(rownames = NA) %>% rownames_to_column("gene")
counts_cpm <- cpm(dgl_raw) %>% as_tibble(rownames = NA) %>% rownames_to_column("gene")

counts_raw <- counts_raw %>% rowwise() %>% mutate(mean = mean(c_across(-gene))) %>% arrange(-mean) 
counts_cpm <- counts_cpm %>% rowwise() %>% mutate(mean = mean(c_across(-gene))) %>% arrange(-mean) 

counts_raw %>% 
  select(-gene, -mean) %>% 
  colSums() %>% 
  as_tibble(rownames = NA) %>% 
  rownames_to_column("library") %>% 
  select(library, total_raw = value) -> counts_raw_bylib

counts_cpm %>% 
  select(-gene, -mean) %>% 
  colSums() %>% 
  as_tibble(rownames = NA) %>% 
  rownames_to_column("library") %>% 
  mutate(
    edgr_cpm = value,
#   cpm = round(value, 0),
    mapped = value/1e6,
    `mapped_%` = mapped*100
  ) %>% 
  select(-value) -> counts_cpm_bylib

df_samples %>% 
  select(library = sample, group, `library size` = lib.size) %>%
  left_join(counts_cpm_bylib, .) %>% 
  left_join(counts_raw_bylib) %>% 
  mutate(
    manual_cpm = total_raw / `library size` * 1e6) -> counts_cpm_summary

# counts_cpm_summary %>%
#   ggplot(aes(x = edgr_cpm, y = manual_cpm)) +
#     geom_point() +
#     coord_fixed() +
#     geom_abline(intercept = 0, gradient = 1) +
#     theme_miR()

counts_cpm_summary %>% 
  ggplot(aes(x = group, y = `mapped_%`)) +
  geom_jitter(alpha = 0.6, width = 0.1) +
  ylim(0, NA) +
  ylab('mapped reads \n(% of total reads)') +
  theme_miR_horizontal() 

```

Raw counts for 10 most highly-expressed genes in first 8 samples: 

```{r print_raw, eval = F}
counts_raw[,1:9] %>% head(n = 10) %>% knitr::kable()
```


Same counts expressed as cpm:

```{r print_cpm, eval = F}
counts_cpm[,1:9] %>% head(n = 10) %>% knitr::kable(digits = 0)

```


Distribution of gene abundance:

```{r plot_cpm, eval = F}

library(scales)

counts_cpm %>% 
  filter(mean > 0) %>% 
    ggplot(aes(x = mean)) + 
      geom_histogram(alpha = 0.6, fill = "blue") + 
      scale_x_log10(labels = function(x) sprintf("%g", x)) + 
      ylab("number of genes") +
      xlab("cpm (mean across all libraries)") +
#      geom_vline(aes(xintercept = cpm_threshold), colour = "red", linetype = 2) +
      theme_miR()

```

-->

Filter out any genes that are not expressed at high abundance (using default edgeR parameters).  

```{r filter}

## option 1 = manual
#
# dgl <- dgl[rowSums(cpm(dgl) >= cpm_threshold) >= n_samples, , keep.lib.sizes=FALSE]

## option 2 = HTSFilter
#
# dgl_filter <- HTSFilter(dgl_raw, normalization = "TMM", plot = FALSE)
# dgl <- dgl_filter$filteredData

dgl_keep <- filterByExpr(dgl_raw)     # this accesses group (or can supply a different design)
dgl <- dgl_raw[dgl_keep, , keep.lib.sizes=FALSE]

libsize_raw <- dgl_raw$samples$lib.size %>% median()
libsize_filtered <- dgl$samples$lib.size %>% median()
s_manual <- libsize_raw / 10^5

```

<!--

The optimal fitering threshold, identified by HTSFilter was `r # dgl_filter$s`.  The rule-of-thumb filtering threshold (equating to absolute read count of 10) was `r # s_manual`.  

-->

Median library size before filtering = `r format(libsize_raw, scientific = F, big.mark = ",", digits = 0)`; and after filtering = `r format(libsize_filtered, scientific = F, big.mark = ",", digits = 0)`.  

There are `r nrow(dgl_raw)` genes in the raw dataset and `r nrow(dgl)` genes in the filtered dataset.    


```{r filter_edgeR_manual, include=FALSE}

# keep <- filterByExpr(dgl)
# dgl <- dgl[keep, , keep.lib.sizes=FALSE]
# 
# print(paste0("number of rows = ", nrow(dgl)))
# dgl$samples %>% knitr::kable()

```


<br>
<br>

Normalisation method = `r norm_method`.  (Either trimmed median of M-values method or using spike-in.)    

Top rows after normalisation (note normalisation factors):  

```{r normalise_TMM}

if (norm_method == "TMM") {

  # normalise reads
  dgl_raw <- calcNormFactors(dgl_raw, method = "TMM")
  dgl <- calcNormFactors(dgl, method = "TMM")
  dgl$samples %>% head() %>% knitr::kable()
  
}

```


<!-- Or normalise to spike-in.  Do not do both!  -->

```{r normalise_spike}

if (norm_method == "spike") {

  # normalise reads
  norm.factors <- spike_in_factor / dgl$samples$lib.size
  norm.factors <- norm.factors / prod(norm.factors)^(1/length(norm.factors))
  dgl$samples$norm.factors <- norm.factors
  
  dgl$samples %>% head() %>% knitr::kable()

}

```


<br>
<br>
Library size profiles after filtering and normalisation:  

```{r plot_libsize}

library_plot_density(
  cpm_raw = dgl_raw %>% cpm(normalized.lib.sizes = TRUE),
  cpm_filtered = dgl %>% cpm(normalized.lib.sizes = TRUE)
)

dgl %>% 
  cpm(normalized.lib.sizes = FALSE) %>% 
  library_plot() + ggtitle("Libraries before normalisation")

dgl %>% 
  cpm(normalized.lib.sizes = TRUE) %>% 
  library_plot() + ggtitle("Libraries after normalisation")

```

<br>
<br>

## Unbiased exploration

### MDS (multi-dimensional scaling plot)

```{r MDS}

plotMDS(
  dgl, 
  gene.selection = 'pairwise',
  col = as.integer(dgl$samples$group), # colour
  labels = dgl$samples$library 
)

```

<br>
<br>

### Principle component analysis (PCA)

```{r PCA}

pca <- pca_plot_RWH(dgl, labels = 'library', color = 'group', 'PC1','PC2') 
pca$scree_plot
pca$pca_plot + sc_miR_groups

pca <- pca_plot_RWH(dgl, labels = 'library', color = 'group', 'PC1','PC3') 
pca$pca_plot + sc_miR_groups

pca <- pca_plot_RWH(dgl, labels = 'library', color = 'group', 'PC2','PC3') 
pca$pca_plot + sc_miR_groups


```


Significance testing:

```{r pca_pvals}

pca_pvals(pca$pca, c('group'), dgl) %>% knitr::kable()

```

```{r save_dgl}

list(
  df_samples = df_samples,
  contrast_name = contrast_name,
  dgl = dgl) %>% saveRDS(file = here(dir_save, "dea_objects.rds"))

```

